sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1070 @@
+                await handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                await handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                await handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: %s", parsed_message.kind)
+
+        async def on_close(close_status_code: int, close_msg: Optional[str] = None):
+            if self._state == WebPubSubClientState.CONNECTED:
+                _LOGGER.info(
+                    "WebSocket connection closed. Code: %s, Reason: %s",
+                    close_status_code,
+                    close_msg,
+                )
+
+                self._last_close_event = CloseEvent(close_status_code=close_status_code, close_reason=close_msg)
+                # clean ack cache
+                self._ack_map.clear()
+
+                if self._is_stopping:
+                    _LOGGER.warning("The client is stopping state. Stop recovery.")
+                    await self._handle_connection_close_and_no_recovery()
+                    return
+
+                if self._last_close_event and self._last_close_event.close_status_code == 1008:
+                    _LOGGER.warning("The websocket close with status code 1008. Stop recovery.")
+                    await self._handle_connection_close_and_no_recovery()
+                    return
+
+                if not self._protocol.is_reliable_sub_protocol:
+                    _LOGGER.warning("The protocol is not reliable, recovery is not applicable")
+                    await self._handle_connection_close_and_no_recovery()
+                    return
+
+                recovery_url = self._build_recovery_url()
+                if not recovery_url:
+                    _LOGGER.warning("Connection id or reconnection token is not available")
+                    await self._handle_connection_close_and_no_recovery()
+                    return
+
+                self._state = WebPubSubClientState.RECOVERING
+                recovery_start = time.time()
+                while (time.time() - recovery_start < _RECOVERY_TIMEOUT) and not self._is_stopping:
+                    try:
+                        await self._connect(recovery_url)
+                        _LOGGER.debug("Recovery succeeded")
+                        return
+                    except:  # pylint: disable=bare-except
+                        _LOGGER.debug("Try to recover after %d seconds", _RECOVERY_RETRY_INTERVAL)
+                        await asyncio.sleep(_RECOVERY_RETRY_INTERVAL)
+
+                _LOGGER.warning("Recovery attempts failed after 30 seconds or the client is stopping")
+                await self._handle_connection_close_and_no_recovery()
+
+        if self._is_stopping:
+            raise OpenClientError("Can't open a client during closing")
+        await self._wait_for_tasks([self._task_seq_ack])
```

Comment:
will the previous run_forever task need to be stopped here before re-start?

Reply:
There are two background task: run_forever task and seq_ack task. The latter is to update sequenceId which is for recovery and here is detail info https://github.com/Azure/azure-webpubsub/blob/main/protocols/client/client-spec.md#32-sequenceid.
There is one case that users may reopen the client by registering STOPPED callback (e.g. this [testcase](https://github.com/Azure/azure-sdk-for-python/blob/a309d8fdd6d8b95a792e8f33d5920fa1f42bfcd5/sdk/webpubsub/azure-messaging-webpubsubclient/tests/test_smoke.py#L57-L76), then we need to wait for it done before recreate new one.